src/library_syscall.js

```diff
@@ -1081,11 +1081,17 @@ var SyscallsLibrary = {
     if (!stream.getdents) {
       stream.getdents = FS.readdir(stream.path);
     }
+
+    var struct_size = {{{ C_STRUCTS.dirent.__size__ }}};
     var pos = 0;
-    while (stream.getdents.length > 0 && pos + {{{ C_STRUCTS.dirent.__size__ }}} <= count) {
+    var off = FS.llseek(stream, 0, {{{ cDefine('SEEK_CUR') }}});
+
+    var idx = Math.floor(off / struct_size);
+
+    while (idx < stream.getdents.length && pos + struct_size <= count) {
       var id;
       var type;
-      var name = stream.getdents.pop();
+      var name = stream.getdents[stream.getdents.length - 1 - idx];
```

Comment:
Hmm. .are we reading these backwards?  Why not just `stream.getdents[idx]`?

Reply:
The old code iterated backwards by using `.pop()` this just preserves that behavior. I can put a comment there to make that clear.

Reply:
if we switch to forward iteration presumably we can avoid the change in `lib/fetch/asmfs.cpp`?   Do we have any tests that break?  (i.e. rely on the order iteration?)

Reply:
The `d_off` change in `asmfs.cpp` is not related to iteration order, that's just how `readdir()` works. The tests that depend on iteration order are `other.test_mkdir_silly` and `other.test_readdir_r_silly`

Reply:
(I'm open to changing those tests for the purposes of changing iteration order)

Reply:
I think its worth a try.